## Machine learning Model Development

- In this section, we will develop and evaluate machine learning models to predict life expectancy based on various public health metrics, environmental factors, and demographic data. The process will be carried out in a systematic manner, starting from data preparation, model training, and evaluation, followed by model tuning and final evaluation.




In [8]:
# importing libraries and prepared dataset

import pandas as pd
from sklearn.model_selection import train_test_split

data = "/Users/alexandreribeiro/Documents/GitHub/final_project/notebooks/dataset_for_ml.csv"

df = pd.read_csv(data)

df.sample(5)

population_city  greenspacearea_km2       AQI  adjusted_obesity_rate  \
195          1.768860            0.827357 -0.614014               2.072841   
1794        -0.405541            2.261801 -2.092229               1.224708   
339          1.094393            0.212659  1.338461               0.634357   
601          0.546686            0.410107  2.662695              -1.131691   
977          0.087315            0.008053 -0.213665               0.073358   

      adjusted_smoking_rate  adjusted_copd_rate  adjusted_depression_rate  \
195                1.442063            1.678737                  1.243903   
1794               1.272402            1.478700                  1.089777   
339                0.925072            0.656073                  0.627727   
601               -1.123796           -1.119047                 -1.080786   
977               -0.077285           -0.468440                 -0.644293   

      adjusted_life_expectancy  
195                   1.710293  
1794                  0.768165  
339                   0.657804  
601                  -1.039814  
977                   0.359496

#### Defining features and target variable

- The first step in developing a machine learning model is to define the features and target variable. In this case, the target variable is the life expectancy, and the features are the various public health metrics, environmental factors, and demographic data. We will use the following features to predict life expectancy.

In [10]:
# Selected features

selected_features = ['population_city', 'greenspacearea_km2', 'AQI', 'adjusted_obesity_rate',
       'adjusted_smoking_rate', 'adjusted_copd_rate',
       'adjusted_depression_rate']

# Define the target variable (y) and the feature set (X)

y = df['adjusted_life_expectancy']  # Target variable
X = df[selected_features]  # Using the selected features

# Display the shapes of X and y

X.shape, y.shape

((3238, 7), (3238,))

#### Spliting the data into training and testing sets

- Split the dataset into training and testing sets to train the model on one portion and test it on another.

- test_size=0.2: This means that 20% of the data will be used for testing, and the remaining 80% will be used for training.
- random_state=42: Setting a random state ensures that the split is reproducible.

In [12]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the training and testing sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (2590, 7)
X_test shape: (648, 7)
y_train shape: (2590,)
y_test shape: (648,)


#### Model Selection:

- We will start with a few common regression models since you’re predicting a continuous target variable (adjusted life expectancy).

We’ll try these models:

- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor
- Gradient Boosting Regressor
- Support Vector Regressor (SVR)

- After selecting the models, we’ll compare their performance to choose the best one.

In [13]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Define the models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the results
results = {}

# Train and evaluate each model
for model_name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results
    results[model_name] = {
        "MAE": mae,
        "MSE": mse,
        "R²": r2
    }

    print(f"{model_name} results:")
    print(f"  - MAE: {mae:.3f}")
    print(f"  - MSE: {mse:.3f}")
    print(f"  - R²: {r2:.3f}\n")

# Convert the results to a DataFrame for easier comparison
import pandas as pd
results_df = pd.DataFrame(results).T
print(results_df)

Linear Regression results:
  - MAE: 0.149
  - MSE: 0.048
  - R²: 0.957

Decision Tree results:
  - MAE: 0.065
  - MSE: 0.026
  - R²: 0.976

Random Forest results:
  - MAE: 0.065
  - MSE: 0.017
  - R²: 0.985

Gradient Boosting results:
  - MAE: 0.121
  - MSE: 0.031
  - R²: 0.972

SVR results:
  - MAE: 0.125
  - MSE: 0.039
  - R²: 0.965

                        MAE       MSE        R²
Linear Regression  0.148989  0.047983  0.956544
Decision Tree      0.065243  0.026446  0.976049
Random Forest      0.064775  0.016573  0.984990
Gradient Boosting  0.121096  0.030596  0.972290
SVR                0.125336  0.039029  0.964653


In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import pandas as pd

# Define the models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the results
results = []

# Train and evaluate each model
for model_name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results in the list
    results.append({
        "Model": model_name,
        "MAE": mae,
        "MSE": mse,
        "R²": r2
    })

# Convert the results to a DataFrame for easier comparison
results_df = pd.DataFrame(results)
results_df

Model       MAE       MSE        R²
0  Linear Regression  0.148989  0.047983  0.956544
1      Decision Tree  0.064775  0.026647  0.975867
2      Random Forest  0.064788  0.016654  0.984917
3  Gradient Boosting  0.121070  0.030654  0.972238
4                SVR  0.125336  0.039029  0.964653

#### Cross-validation:

- We will use cross-validation to evaluate the performance of the models. Cross-validation is a technique that splits the data into multiple subsets (folds) and trains the model on different combinations of these subsets. This helps to get a more accurate estimate of the model’s performance.

In [16]:
from sklearn.model_selection import cross_val_score
import numpy as np

# Define the model
model = RandomForestRegressor()  # You can choose the model you prefer

# Perform 5-fold cross-validation
cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')

# Output the cross-validation results
print("Cross-Validation R² Scores: ", cv_scores)
print("Mean R²: ", np.mean(cv_scores))
print("Standard Deviation of R²: ", np.std(cv_scores))

Cross-Validation R² Scores:  [0.98399701 0.98101546 0.98438577 0.98448617 0.98764831]
Mean R²:  0.9843065439930305
Standard Deviation of R²:  0.0021039479067733115


#### Results:

- Cross-Validation R² Scores: The individual R² scores across the 5 folds are all very high, indicating that the model is consistently performing well on different subsets of the data.

- Mean R²: The average R² score across the 5 folds is approximately 0.984, which suggests that the model explains about 98.4% of the variance in the target variable.

- Standard Deviation of R²: The standard deviation is very low (0.0021), indicating that the model’s performance is consistent across different data splits. This is a good sign as it shows that the model is not overly dependent on any particular subset of the data.


In [19]:
# Lets check the remaining models

# List of models to evaluate

models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the cross-validation results
cv_results = {}

# Perform cross-validation for each model
for model_name, model in models.items():
    # Perform cross-validation with 5 folds
    cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')
    
    # Store the mean and standard deviation of R² scores
    cv_results[model_name] = {
        "Mean R²": cv_scores.mean(),
        "Std Dev R²": cv_scores.std()
    }

# Convert the results to a DataFrame for easier comparison
cv_results_df = pd.DataFrame(cv_results).T
cv_results_df

Mean R²  Std Dev R²
Linear Regression  0.953557    0.004549
Decision Tree      0.973028    0.002493
Random Forest      0.984303    0.002198
Gradient Boosting  0.968400    0.001951
SVR                0.960934    0.004133

#### Using Hyperparameter Tuning to try to improve the model performance

- Using GridSearchCV to find the best hyperparameters for the model. GridSearchCV is a technique that searches for the best combination of hyperparameters by evaluating the model’s performance on different combinations of hyperparameters.

In [20]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

# Define the parameter grid for Random Forest
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'max_features': ['auto', 'sqrt', 'log2'],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize the Random Forest Regressor
rf = RandomForestRegressor()

# Initialize GridSearchCV with 5-fold cross-validation
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, 
                              cv=5, scoring='r2', n_jobs=-1, verbose=2)

# Fit the GridSearchCV on the training data
grid_search_rf.fit(X_train, y_train)

# Best parameters and the corresponding score
best_params_rf = grid_search_rf.best_params_
best_score_rf = grid_search_rf.best_score_

print(f"Best Parameters for Random Forest: {best_params_rf}")
print(f"Best R² Score for Random Forest: {best_score_rf:.4f}")

Fitting 5 folds for each of 324 candidates, totalling 1620 fits
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s

[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=

/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.4s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=5, n_estimators=300; total time=   1.1s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.4s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=5, n_estimators=100; total time=   0.4s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, n_estimators=200; total time=   0.8s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=300; total time=   1.0s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.3s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, n_estimators=200; total time=   0.6s
[CV] END max_depth=No

/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:540: FitFailedWarning: 
540 fits failed out of a total of 1620.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
335 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/base.py", line 1466, in wrapper
    estimator._validate_params()
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/base.py", line 666, in _validate_params
    validate_parameter_

Best Parameters for Random Forest: {'max_depth': 20, 'max_features': 'log2', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}
Best R² Score for Random Forest: 0.9763


In [21]:
# Get the best model from GridSearchCV
best_rf = grid_search_rf.best_estimator_

# Make predictions on the test set
y_pred_rf = best_rf.predict(X_test)

# Evaluate the performance
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"Random Forest - Tuned Model Performance:")
print(f"  - MAE: {mae_rf:.3f}")
print(f"  - MSE: {mse_rf:.3f}")
print(f"  - R²: {r2_rf:.3f}")

Random Forest - Tuned Model Performance:
  - MAE: 0.094
  - MSE: 0.025
  - R²: 0.977


#### Overfitting check

- Compare the performance of the model on the training data versus the test data. A common method is to look at the difference between training and validation scores, especially during cross-validation.

In [25]:
# Assuming you've already done hyperparameter tuning and have the best model
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Train the best model with the entire training set
best_rf_model = RandomForestRegressor(
    max_depth=20, 
    max_features='log2', 
    min_samples_leaf=1, 
    min_samples_split=2, 
    n_estimators=300,
    random_state=42
)

best_rf_model.fit(X_train, y_train)

# Evaluate the model on the training data
y_train_pred = best_rf_model.predict(X_train)

# Calculate metrics on training data
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)

print("Random Forest - Training Data Performance:")
print(f"  - MAE: {train_mae:.3f}")
print(f"  - MSE: {train_mse:.3f}")
print(f"  - R²: {train_r2:.3f}\n")

Random Forest - Training Data Performance:
  - MAE: 0.033
  - MSE: 0.003
  - R²: 0.997



In [26]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Define the best model with the optimal hyperparameters
best_rf_model = RandomForestRegressor(
    max_depth=20, 
    max_features='log2', 
    min_samples_leaf=1, 
    min_samples_split=2, 
    n_estimators=300,
    random_state=42
)

# Train the model on the entire training set
best_rf_model.fit(X_train, y_train)

# Evaluate the model on the training data
y_train_pred = best_rf_model.predict(X_train)

# Calculate metrics on training data
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)

print("Random Forest - Training Data Performance:")
print(f"  - MAE: {train_mae:.3f}")
print(f"  - MSE: {train_mse:.3f}")
print(f"  - R²: {train_r2:.3f}\n")

# Evaluate the model on the test data
y_test_pred = best_rf_model.predict(X_test)

# Calculate metrics on test data
test_mae = mean_absolute_error(y_test, y_test_pred)
test_mse = mean_squared_error(y_test, y_test_pred)
test_r2 = r2_score(y_test, y_test_pred)

print("Random Forest - Test Data Performance:")
print(f"  - MAE: {test_mae:.3f}")
print(f"  - MSE: {test_mse:.3f}")
print(f"  - R²: {test_r2:.3f}\n")

Random Forest - Training Data Performance:
  - MAE: 0.033
  - MSE: 0.003
  - R²: 0.997

Random Forest - Test Data Performance:
  - MAE: 0.093
  - MSE: 0.025
  - R²: 0.977



#### Conclusion:

- A difference of 0.02 in R² is not a strong indication of overfitting. It’s more likely a sign that the model is performing well. Overfitting concerns arise more when there are larger discrepancies between training and test performance. In this case the model seems to be well-tuned, and the small difference are likely acceptable. 

#### Trying XGBoost Regressor

- XGBoost is a powerful and efficient implementation of the gradient boosting algorithm. It is known for its speed and performance, making it a popular choice for machine learning tasks. We will use the XGBoost Regressor to predict life expectancy and evaluate its performance.

In [29]:
import xgboost as xgb
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV

# Initialize XGBoost model

xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

# Train the model

xgb_model.fit(X_train, y_train)

# Make predictions on the test set

y_pred = xgb_model.predict(X_test)

# Evaluate the model

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Display the evaluation metrics in a dataframe

results = pd.DataFrame(data=[mae, mse, r2], index=['MAE', 'MSE', 'R²'], columns=['XGBoost'])

results



XGBoost
MAE  0.066697
MSE  0.014045
R²   0.987280

#### Hyperparameter Tuning for XGBoost Regressor

In [31]:
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.3],
    'subsample': [0.7, 0.8, 0.9],
    'colsample_bytree': [0.7, 0.8, 0.9]
}

grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, cv=5, scoring='r2', n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

print("Best Parameters for XGBoost:", grid_search.best_params_)
print("Best R² Score for XGBoost:", grid_search.best_score_)

# Evaluate the best model on the test set
best_xgb_model = grid_search.best_estimator_
y_pred_best = best_xgb_model.predict(X_test)

# Calculate metrics on test data
best_mae = mean_absolute_error(y_test, y_pred_best)
best_mse = mean_squared_error(y_test, y_pred_best)
best_r2 = r2_score(y_test, y_pred_best)

# Display the evaluation metrics in a dataframe

results['XGBoost Tuned'] = [best_mae, best_mse, best_r2]

results

Fitting 5 folds for each of 243 candidates, totalling 1215 fits
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.8; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.8; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.8; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END

XGBoost  XGBoost Tuned
MAE  0.066697       0.059276
MSE  0.014045       0.012146
R²   0.987280       0.989000

##### Results:

- MAE (Mean Absolute Error) decreased from 0.0667 to 0.0593, indicating that the average error between the predicted and actual values has reduced.

- MSE (Mean Squared Error) also decreased from 0.0140 to 0.0121, which means that the model’s overall error has been reduced, especially penalizing larger errors.

- R² (Coefficient of Determination) increased slightly from 0.9873 to 0.9890, showing that the model now explains slightly more variance in the data.

#### Checking for overfitting on XGBoost Regressor

In [33]:
import xgboost as xgb

# Best parameters from your hyperparameter tuning
best_params_xgb = {
    'colsample_bytree': 0.9,
    'learning_rate': 0.1,
    'max_depth': 7,
    'n_estimators': 300,
    'subsample': 0.9
}

# Initialize and train the XGBoost model with the best parameters
tuned_xgb_model = xgb.XGBRegressor(**best_params_xgb)
tuned_xgb_model.fit(X_train, y_train)

# Now, evaluate the tuned XGBoost model on both the training and test data
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Evaluate on the training data
y_train_pred_xgb = tuned_xgb_model.predict(X_train)
train_mae_xgb = mean_absolute_error(y_train, y_train_pred_xgb)
train_mse_xgb = mean_squared_error(y_train, y_train_pred_xgb)
train_r2_xgb = r2_score(y_train, y_train_pred_xgb)

print("Tuned XGBoost - Training Data Performance:")
print(f"  - MAE: {train_mae_xgb:.3f}")
print(f"  - MSE: {train_mse_xgb:.3f}")
print(f"  - R²: {train_r2_xgb:.3f}\n")

# Evaluate on the test data
y_test_pred_xgb = tuned_xgb_model.predict(X_test)
test_mae_xgb = mean_absolute_error(y_test, y_test_pred_xgb)
test_mse_xgb = mean_squared_error(y_test, y_test_pred_xgb)
test_r2_xgb = r2_score(y_test, y_test_pred_xgb)

print("Tuned XGBoost - Test Data Performance:")
print(f"  - MAE: {test_mae_xgb:.3f}")
print(f"  - MSE: {test_mse_xgb:.3f}")
print(f"  - R²: {test_r2_xgb:.3f}")

Tuned XGBoost - Training Data Performance:
  - MAE: 0.005
  - MSE: 0.000
  - R²: 1.000

Tuned XGBoost - Test Data Performance:
  - MAE: 0.062
  - MSE: 0.013
  - R²: 0.988
